packages/definitions-parser/src/packages.ts

```diff
@@ -83,71 +107,109 @@ export class AllPackages {
-    const versions = this.data.get(typesDirectoryName);
+    const versions = await this.tryGetTypingsVersions(typesDirectoryName);
     return (
       versions && versions.tryGet(new semver.Range(pkg.version === "*" ? "*" : `^${formatTypingVersion(pkg.version)}`))
     );
   }
 
-  allPackages(): readonly AnyPackage[] {
-    return [...this.allTypings(), ...this.allNotNeeded()];
+  async tryGetTypingsVersions(typesDirectoryName: string): Promise<TypingsVersions | undefined> {
+    let versions = this.types.get(typesDirectoryName);
+    if (versions) {
+      return versions;
+    }
+    if (this.errors.has(typesDirectoryName)) {
+      return undefined;
+    }
+    const raw = await getTypingInfo(typesDirectoryName, this.dt);
+    if (!raw) {
+      return undefined;
+    }
+    if ("errors" in raw) {
+      this.errors.set(typesDirectoryName, raw.errors);
+      return undefined;
+    }
+    versions = new TypingsVersions(this.dt, raw, this.moduleResolutionHost);
+    this.types.set(typesDirectoryName, versions);
+    return versions;
+  }
+
+  async allPackages(): Promise<readonly AnyPackage[]> {
+    return [...(await this.allTypings()), ...this.allNotNeeded()];
   }
 
   /** Note: this includes older version directories (`foo/v0`) */
-  allTypings(): readonly TypingsData[] {
-    return assertSorted(Array.from(flattenData(this.data)), (t) => t.name);
+  async allTypings(): Promise<readonly TypingsData[]> {
+    await this.readAllTypings();
+    return Array.from(flattenData(this.types));
   }
 
-  allLatestTypings(): readonly TypingsData[] {
-    return assertSorted(
-      Array.from(this.data.values()).map((versions) => versions.getLatest()),
-      (t) => t.name
-    );
+  async allLatestTypings(): Promise<readonly TypingsData[]> {
+    await this.readAllTypings();
+    return Array.from(this.types.values()).map((versions) => versions.getLatest());
   }
 
   allNotNeeded(): readonly NotNeededPackage[] {
     return this.notNeeded;
   }
 
   /** Returns all of the dependencies *that are typed on DT*, ignoring others, and including test dependencies. */
-  *allDependencyTypings(pkg: TypingsData): Iterable<TypingsData> {
+  async *allDependencyTypings(pkg: TypingsData): AsyncIterable<TypingsData> {
     for (const [name, version] of pkg.allPackageJsonDependencies()) {
       if (!name.startsWith(`@${scopeName}/`)) continue;
       if (pkg.name === name) continue;
       const typesDirectoryName = removeTypesScope(name);
-      const versions = this.data.get(typesDirectoryName);
+      const versions = await this.tryGetTypingsVersions(typesDirectoryName);
       if (versions) {
         yield versions.get(new semver.Range(version), pkg.name + ":" + JSON.stringify((versions as any).versions));
       }
     }
   }
+
+  private async readAllTypings() {
+    if (this.isComplete) {
+      return;
+    }
+    // populate cache so every directory doesn't try to request this from GH
+    await getAllowedPackageJsonDependencies();
```

Comment:
how does this populate a cache? I'm not seeing it in getAllowedPackageJsonDependencies

Reply:
oh I missed the `withCache` wrapping the whole thing